### DISCUSSION TITLE
Why my gripper can not close when I want to grasp the cube?

### DISCUSSION THREAD

**Discussion by Rabibabo** [isaac-sim/IsaacLab#2651] [Q&A] [answered]
I am very confused. When there is no object between my two fingers, the claw can close normally, but when a block is added, the claw does not come into contact with the block and seems to be stuck.
My code as follow
```python
import argparse
from scipy.spatial.transform import Rotation as R
import numpy as np
import copy
import time

from isaaclab.app import AppLauncher

# add argparse arguments
parser = argparse.ArgumentParser(description="Tutorial on using the differential IK controller.")
# parser.add_argument("--robot", type=str, default="franka_panda", help="Name of the robot.")
parser.add_argument("--robot", type=str, default="ur10", help="Name of the robot.")
parser.add_argument("--num_envs", type=int, default=1, help="Number of environments to spawn.")
# append AppLauncher cli args
AppLauncher.add_app_launcher_args(parser)
# parse the arguments
args_cli = parser.parse_args()

# launch omniverse app
app_launcher = AppLauncher(args_cli)
simulation_app = app_launcher.app

"""Rest everything follows."""

import torch
# from omni.physx.scripting import create_material, apply_material_to_prim
import isaaclab.sim as sim_utils
# from isaaclab.sim.spawners import materials
from isaaclab.assets import AssetBaseCfg, RigidObjectCfg
from isaaclab.controllers import DifferentialIKController, DifferentialIKControllerCfg
from isaaclab.managers import SceneEntityCfg
from isaaclab.markers import VisualizationMarkers
from isaaclab.markers.config import FRAME_MARKER_CFG
from isaaclab.scene import InteractiveScene, InteractiveSceneCfg
from isaaclab.utils import configclass
from isaaclab.utils.assets import ISAAC_NUCLEUS_DIR
from isaaclab.utils.math import subtract_frame_transforms

# ##
# # Pre-defined configs
# ##
from isaaclab_assets import FRANKA_PANDA_HIGH_PD_CFG, UR10_CFG, UR5E_DH_CFG, RoBotiq2FGripper_CFG, DH_PGC140_Gripper_CFG # isort:skip


# def get_init_rot():
#     """Get the initial rotation of the robot."""
#     x,y,z,w = R.from_euler("xyz", [0, 0, 0], degrees=True).as_quat()
#     quat = (w,x,y,z)
#     return quat


@configclass
class TableTopSceneCfg(InteractiveSceneCfg):
    """Configuration for a cart-pole scene."""

    # ground plane
    ground = AssetBaseCfg(
        prim_path="/World/defaultGroundPlane",
        spawn=sim_utils.GroundPlaneCfg(),
        init_state=AssetBaseCfg.InitialStateCfg(pos=(0.0, 0.0, -1.05)),
    )

    # lights
    dome_light = AssetBaseCfg(
        prim_path="/World/Light", spawn=sim_utils.DomeLightCfg(intensity=3000.0, color=(0.75, 0.75, 0.75))
    )

    # mount
    table = AssetBaseCfg(
        prim_path="{ENV_REGEX_NS}/Table",
        spawn=sim_utils.UsdFileCfg(
            usd_path=f"{ISAAC_NUCLEUS_DIR}/Props/Mounts/SeattleLabTable/table_instanceable.usd", scale=(1.0, 1.0, 1.0)
        ),
    )

    cube = RigidObjectCfg(
        prim_path="{ENV_REGEX_NS}/Cube",
        spawn=sim_utils.CuboidCfg(
            size = (0.03, 0.03, 0.03),
            rigid_props=sim_utils.RigidBodyPropertiesCfg(),
            mass_props=sim_utils.MassPropertiesCfg(mass=0.2),
            collision_props=sim_utils.CollisionPropertiesCfg(
                collision_enabled=True,
                contact_offset=1e-3,
                rest_offset=0.0,
                torsional_patch_radius=1e-3,
                min_torsional_patch_radius=1e-3
            ),
            physics_material = sim_utils.spawners.RigidBodyMaterialCfg(
                static_friction=1.5,
                dynamic_friction=1.5,
                restitution=0.0,
            ),
            visual_material=sim_utils.PreviewSurfaceCfg(diffuse_color=(1.0, 1.0, 0.0), metallic=0.2),
        ),
        init_state=RigidObjectCfg.InitialStateCfg(
            pos = (0.0, 0.0, 0.0),
        ),
    )
    robot = DH_PGC140_Gripper_CFG.replace(prim_path="{ENV_REGEX_NS}/Robot") # type: ignore
    robot.init_state.pos = (0.0, 0.0, 0.15)
    robot.init_state.rot = (0.0, 1.0, 0.0, 0.0)


def run_simulator(sim: sim_utils.SimulationContext, scene: InteractiveScene):
    """Runs the simulation loop."""
    # Extract scene entities
    # note: we only do this here for readability.
    count = 0
    sim_dt = sim.get_physics_dt()
    robot = scene["robot"]
    # robot_entity_cfg = SceneEntityCfg("robot", joint_names=['finger_joint', 'left_inner_knuckle_joint', 'right_inner_knuckle_joint', 'right_outer_knuckle_joint', 'left_inner_finger_joint', 'right_inner_finger_joint'])
    # Simulation loop
    
    while simulation_app.is_running():
        if count % 500 == 0:
            # reset counter
            count = 0
            root_state = robot.data.default_root_state.clone()
            root_state[:, :3] += scene.env_origins
            robot.write_root_pose_to_sim(root_state[:, :7])
            robot.write_root_velocity_to_sim(root_state[:, 7:])
            # set joint positions with some noise
            joint_pos, joint_vel = robot.data.default_joint_pos.clone(), robot.data.default_joint_vel.clone()
            joint_pos += torch.zeros_like(joint_pos)
            robot.write_joint_state_to_sim(joint_pos, joint_vel)
            # clear internal buffers
            scene.reset()
            print("[INFO]: Resetting robot state...")

        joint_pose = torch.tensor([[0.025,0.025]], device=robot.device)
        robot.set_joint_position_target(joint_pose)
        robot.write_data_to_sim()
        count += 1
        sim.step()


def main():
    """Main function."""
    # Load kit helper
    sim_cfg = sim_utils.SimulationCfg(dt=0.01, device=args_cli.device)
    sim = sim_utils.SimulationContext(sim_cfg)
    # Set main camera
    sim.set_camera_view([2.5, 2.5, 2.5], [0.0, 0.0, 0.0]) # type: ignore
    # Design scene
    scene_cfg = TableTopSceneCfg(num_envs=args_cli.num_envs, env_spacing=2.0)
    scene = InteractiveScene(scene_cfg)
    
    # Play the simulator
    sim.reset()
    # Now we are ready!
    print("[INFO]: Setup complete...")
    # Run the simulator
    run_simulator(sim, scene)


def goal_pose_callback(data):
    print("Received goal pose")


if __name__ == "__main__":
    # run the main function
    main()
    # close sim app
    simulation_app.close()
```

with out obj the gripper like this

![Image](https://github.com/user-attachments/assets/91388325-8d2b-43bf-8bb5-500869540630)

with obj the gripper like this

![Image](https://github.com/user-attachments/assets/6d6a7f1b-4e63-47db-971d-84d8f546f701)

How can I solve it?

**Comment by RandomOakForest** [comment]
Thanks for posting this. I'll move this to our _Discussions_ section for follow up. In the meantime, here are a few areas to investigate. 

### 1. **Collision Configuration**

- **Collider Types**: Ensure both gripper fingers and cube use appropriate collision meshes (SDF or convex hull)[^1][^2][^3]. The default convex hull approximation might create gaps.

```python
# For cube (already using precise settings)
collision_props=sim_utils.CollisionPropertiesCfg(
    collision_enabled=True,
    contact_offset=1e-3,
    rest_offset=0.0
)
```

- **Contact Offset**: Increase contact offset for both gripper and cube to 0.005-0.01m to ensure early collision detection[^3].


### 2. **Physics Materials**

- Verify friction coefficients match between gripper and cube:

```python
# Current cube settings (adequate)
physics_material = sim_utils.RigidBodyMaterialCfg(
    static_friction=1.5,
    dynamic_friction=1.5
)
```

- Ensure gripper fingers have equivalent or higher friction values in their USD configuration[^4][^5].


### 3. **Joint Control Parameters**

The position control approach might be insufficient against resistance:

```python
# Current position control
joint_pose = torch.tensor([[0.025,0.025]])
robot.set_joint_position_target(joint_pose)
```

Try switching to **velocity control** with force limits:

```python
target_velocity = torch.tensor([[0.5, 0.5]])  # m/s
robot.set_joint_velocity_target(target_velocity)
```


### 4. **Simulation Tuning**

Adjust physics parameters in `SimulationCfg`:

```python
sim_cfg = sim_utils.SimulationCfg(
    dt=0.005,  # Reduced timestep
    substeps=4,  # Increased substeps
    solver_iterations=64  # Higher iteration count
)
```


### 5. **Gripper Configuration**

Ensure proper joint setup in the gripper USD file:

- Confirm all finger joints have **collision meshes** enabled[^6][^7]
- Verify **mimic joint relationships** are properly configured[^7]
- Check joint drive parameters (stiffness/damping)[^8][^3]:

```python
# Example from search results
stiffness=0.0,
damping=5000.0,
max_force=180.0
```


### 6. **Debugging Tools**

Add contact sensors to visualize interactions[^9][^10]:

```python
from isaacsim.sensors.physics import ContactSensor

sensor = ContactSensor(
    prim_path="/World/Robot/finger/Contact_Sensor",
    frequency=60,
    min_threshold=0.1
)
```


### Recommended Fix Sequence

1. First verify collision meshes using **View > Physics > Colliders**
2. Increase contact offset values for both objects
3. Switch to velocity-controlled gripper motion
4. Gradually increase max force/torque limits
5. Add contact sensors to validate actual collisions

If issues persist after these adjustments, consider sharing a USD capture using OmniPVD for deeper physics debugging[^2][^4]. 

_References_

[^1]: https://forums.developer.nvidia.com/t/objects-clipping-through-gripper/253107

[^2]: https://forums.developer.nvidia.com/t/isaac-sim-gripper-going-through-object/291991

[^3]: https://forums.developer.nvidia.com/t/objects-penetrating-each-other-collision-physics-issue-in-nvidia-isaac-sim/269387

[^4]: https://forums.developer.nvidia.com/t/object-gripping-and-picking/291963

[^5]: https://docs.omniverse.nvidia.com/isaacsim/latest/advanced_tutorials/tutorial_grasp_editor.html

[^6]: https://forums.developer.nvidia.com/t/i-want-to-detect-the-collision-between-the-gripper-and-the-building-blocks-using-sdf-mesh-but-in-the-isaac-sim-simulation-interface-no-collision/303308

[^7]: https://docs.omniverse.nvidia.com/isaacsim/latest/advanced_tutorials/tutorial_advanced_rigging_complex_structures.html

[^8]: https://forums.developer.nvidia.com/t/closing-gripper-around-an-object/325967

[^9]: https://docs.isaacsim.omniverse.nvidia.com/4.5.0/sensors/isaacsim_sensors_physics_contact.html

[^10]: https://docs.isaacsim.omniverse.nvidia.com/4.5.0/robot_simulation/ext_isaacsim_robot_surface_gripper.html

[^11]: https://github.com/isaac-sim/IsaacLab/discussions/1721

[^12]: https://tutorial.j3soon.com/robotics/discussions/isaac-lab/

[^13]: https://github.com/isaac-sim/IsaacLab/discussions/2626

[^14]: https://forums.developer.nvidia.com/t/how-to-add-and-control-a-suction-gripper-on-ur-in-isaacsim/250638

[^15]: https://isaac-sim.github.io/IsaacLab/v2.0.1/source/overview/teleop_imitation.html

[^16]: https://forums.developer.nvidia.com/t/how-to-enable-grippers-collision/271951

[^17]: https://forums.developer.nvidia.com/t/simple-way-to-check-for-collision/177947

[^18]: https://docs.robotsfan.com/isaacsim/latest/physics/index.html

[^19]: https://forums.developer.nvidia.com/t/help-diagnosing-why-prismatic-joints-will-not-fully-close-on-robot-iq-hand-e/312455

[^20]: https://forums.developer.nvidia.com/t/tethering-object-to-a-gripper/231020

[^21]: https://github.com/isaac-sim/IsaacLab/issues/939

[^22]: https://www.youtube.com/watch?v=kRAZZ5OPZyM

[^23]: https://www.youtube.com/watch?v=Eb2zuQxOBlY

[^24]: https://autolab.berkeley.edu/assets/publications/media/2022-ICRA-IPC-GraspSim-kim-submission-highres.pdf

[^25]: https://www.youtube.com/watch?v=pMu3mMG7P4E

**Comment by Rabibabo** [comment]
I have finished it, thank you for your help!!

**Reply by Rabibabo** [reply, marked as answer]
The solution is changing the collision meshes, which is mentioned in "https://docs.omniverse.nvidia.com/isaacsim/latest/advanced_tutorials/tutorial_advanced_rigging_complex_structures.html"
 "To visualize the collision meshes, find the eye icon on top of the viewport, and click Show By Type > Physics > Colliders > All. An outlines should show up surrounding any objects that have collision meshes. If you wish to change any collision meshes, select the part of the object associated with that mesh by clicking on it in the viewport, and then in the Physics section of the Property panel, change the Collider Approximation type to Convex Decomposition, or any other type that’s appropriate for your use case. "

**Reply by JohnBrann** [reply]
Not sure if you saw my comment below but I am not sure this is not working for me!
@Rabibabo

**Comment by Rabibabo** [comment]
However, I have another issue: the scene view in my simulation environment with enable_camera enabled is different from the one without enable_camera enabled. After enabling enable_camera, the scene has very high contrast, as shown in the images. Could you please explain why this is happening?
fig 1: enable_camera
![enable_camera](https://github.com/user-attachments/assets/bbd62f50-0cb7-4033-b883-64baa518c3ca)



fig2: without enable_camera
![without_enable_camera](https://github.com/user-attachments/assets/01ce7bed-8910-4df6-a2b0-bee164c8aa50)

**Comment by imendezval** [comment]
Hello! 
So far I am working on a very similar setup, and I got the inverse kinematic solver working for multiple environments - I didnt have this specific problem with the gripper, but I am just letting you know in case we can help each other later throughout our projects :)

![capture 2025-06-16 17 27 11](https://github.com/user-attachments/assets/48a6407f-58ab-47b6-8430-07d570284fb8)

**Reply by Rabibabo** [reply]
OK，very glad to hear that， hope we can help each other。 How can I contact with you？ just in this discussion？

**Comment by JohnBrann** [comment]
Hello I think I am running into a similar issue and cant seem to solve the problem:

I am using Isaac sim 4.5

I am using a free floating gripper in space in order to label grasps. I got the gripper usd file from the isaac sim assets folder. I am able to control the gripper however the issue I am having is with the collider box of the gripper fingers. 

This is what the gripper looks like by default:
![image](https://github.com/user-attachments/assets/e090fc3b-10dc-46cd-b840-60742620699c)

This is what the gripper looks like using setCollider(path, "convexDecomposition") (I have also tried "sdf")
![image](https://github.com/user-attachments/assets/ad6cf407-2eb0-448d-bec8-3ca687197c64)

Using convexDecomposition looks great however, when I attempt the grasp an object, it clearly is still using the triangle shape collision boxes from convex hull, as the object always shoots out. I have slowed down the simulation and it would only be able to grab an object by it finger tips and never comes in contact further down the finger.

In this image both gripper are already in contact with the object. Although it looks like it is using the convexDecomposition approximation, it is using the convexHull approximation collision boxes.
![image](https://github.com/user-attachments/assets/bc382ed0-c82a-45c1-8259-1c36f0b73eec)


No matter what I do I am getting this issue, even after I set the approximations to convexDecomposition. I have also tired applying convexDecomposition directly through the GUI and still get the same results 
This is some of the error message:

2025-06-17 18:56:02 [352,746ms] [Warning] [omni.physx.plugin] PhysicsUSD: Parse collision - triangle mesh collision (approximation None/MeshSimplification) cannot be a part of a dynamic body, falling back to convexHull approximation: /robotiq_gripper/left_gripper. For dynamic collision please use approximations: convex hull, convex decomposition, box, sphere or SDF approximation.
2025-06-17 18:56:02 [352,747ms] [Warning] [omni.physx.plugin] PhysicsUSD: Parse collision - triangle mesh collision (approximation None/MeshSimplification) cannot be a part of a dynamic body, falling back to convexHull approximation: /robotiq_gripper/left_gripper. For dynamic collision please use approximations: convex hull, convex decomposition, box, sphere or SDF approximation.
2025-06-17 18:56:02 [352,747ms] [Warning] [omni.physx.plugin] PhysicsUSD: Parse collision - triangle mesh collision (approximation None/MeshSimplification) cannot be a part of a dynamic body, falling back to convexHull approximation: /robotiq_gripper/right_gripper. For dynamic collision please use approximations: convex hull, convex decomposition, box, sphere or SDF approximation.
2025-06-17 18:56:02 [352,748ms] [Warning] [omni.physx.plugin] PhysicsUSD: Parse collision - triangle mesh collision (approximation None/MeshSimplification) cannot be a part of a dynamic body, falling back to convexHull approximation: /robotiq_gripper/right_gripper. For dynamic collision please use approximations: convex hull, convex decomposition, box, sphere or SDF approximation.
2025-06-17 18:56:02 [352,748ms] [Warning] [omni.physx.plugin] PhysicsUSD: Parse collision - triangle mesh collision (approximation None/MeshSimplification) cannot be a part of a dynamic body, falling back to convexHull approximation: /robotiq_gripper/right_gripper. For dynamic collision please use approximations: convex hull, convex decomposition, box, sphere or SDF approximation.

I can provide more code if necessary. I appreciate any help and guidance!!

**Reply by Rabibabo** [reply]
As you saw in the warning message
“2025-06-17 18:56:02 [352,746ms] [Warning] [omni.physx.plugin] PhysicsUSD: Parse collision - triangle mesh collision (approximation None/MeshSimplification) cannot be a part of a dynamic body, falling back to convexHull ”
the gripper USD you’re using has a collision model set to triangleMesh, which is not supported by IsaacSim. As a result, it automatically falls back to convexHull. Therefore, you need to load the gripper USD file separately in IsaacSim, change its collision model to convexDecomposition, save it, and then use that modified version in your scene.

**Reply by JohnBrann** [reply]
Thank you! I will try this
